teamcity: failed test: TestBootstrapNewStore

The following tests appear to have failed on master (testrace): TestBootstrapNewStore

You may want to check [for open issues](https://github.com/cockroachdb/cockroach/issues?q=is%3Aissue+is%3Aopen+TestBootstrapNewStore).

[#1489746](https://teamcity.cockroachdb.com/viewLog.html?buildId=1489746):


```
TestBootstrapNewStore
...	/usr/local/go/src/runtime/sema.go:510 +0xf9
sync.(*Cond).Wait(0xc00df95b20)
	/usr/local/go/src/sync/cond.go:56 +0x8e
github.com/cockroachdb/cockroach/pkg/storage/engine.(*RocksDB).syncLoop(0xc00df95a00)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/rocksdb.go:693 +0x81
created by github.com/cockroachdb/cockroach/pkg/storage/engine.(*RocksDB).open
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/rocksdb.go:680 +0x4e7

goroutine 58395 [sync.Cond.Wait]:
runtime.goparkunlock(...)
	/usr/local/go/src/runtime/proc.go:307
sync.runtime_notifyListWait(0xc011337150, 0x27)
	/usr/local/go/src/runtime/sema.go:510 +0xf9
sync.(*Cond).Wait(0xc011337140)
	/usr/local/go/src/sync/cond.go:56 +0x8e
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).worker(0xc00eaf2d80, 0x5ba61c0, 0xc00abc42d0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:192 +0x9c
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).Start.func2(0x5ba61c0, 0xc00abc42d0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:161 +0x56
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc00d3979b0, 0xc011ee1e00, 0xc00d3979a0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:196 +0x160
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:189 +0xc4

goroutine 58170 [chan receive]:
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).Start.func1(0x5ba61c0, 0xc00df0aa20)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:151 +0x64
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc00ff3b670, 0xc011ee1e00, 0xc005df8300)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:196 +0x160
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:189 +0xc4

goroutine 58379 [sync.Cond.Wait]:
runtime.goparkunlock(...)
	/usr/local/go/src/runtime/proc.go:307
sync.runtime_notifyListWait(0xc011337150, 0x17)
	/usr/local/go/src/runtime/sema.go:510 +0xf9
sync.(*Cond).Wait(0xc011337140)
	/usr/local/go/src/sync/cond.go:56 +0x8e
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).worker(0xc00eaf2d80, 0x5ba61c0, 0xc0094b6480)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:192 +0x9c
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).Start.func2(0x5ba61c0, 0xc0094b6480)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:161 +0x56
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc00d397740, 0xc011ee1e00, 0xc00d397730)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:196 +0x160
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:189 +0xc4

goroutine 58086 [sync.Cond.Wait]:
runtime.goparkunlock(...)
	/usr/local/go/src/runtime/proc.go:307
sync.runtime_notifyListWait(0xc00202f1d0, 0x2c)
	/usr/local/go/src/runtime/sema.go:510 +0xf9
sync.(*Cond).Wait(0xc00202f1c0)
	/usr/local/go/src/sync/cond.go:56 +0x8e
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).worker(0xc0119523f0, 0x5ba61c0, 0xc00e54b4a0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:192 +0x9c
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).Start.func2(0x5ba61c0, 0xc00e54b4a0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:161 +0x56
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc00fac84d0, 0xc011ee1e00, 0xc00fac84c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:196 +0x160
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:189 +0xc4




```

Please assign, take a look and update the issue accordingly.

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x3fa3156]

goroutine 58434 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc011ee1e00, 0x5ba61c0, 0xc00b467050)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:181 +0x155
panic(0x4951a00, 0x837e440)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/cockroachdb/cockroach/pkg/storage/reports.(*replicationConstraintStatsReportSaver).loadPreviousVersion(0xc00b467110, 0x5ba61c0, 0xc00b467080, 0x0, 0x0, 0xc00b571320, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/reports/constraint_report.go:211 +0xf6
github.com/cockroachdb/cockroach/pkg/storage/reports.(*replicationConstraintStatsReportSaver).Save.func1(0x5ba61c0, 0xc00b467080, 0xc00b571320, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/reports/constraint_report.go:266 +0xbd
github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).Txn.func1(0x5ba61c0, 0xc00b467080, 0xc00b571320, 0xe860932606d3b1ae, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:695 +0x5f
github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).exec(0xc00b571320, 0x5ba61c0, 0xc00b467080, 0xc0118e7010, 0x1, 0xc00b571320)
	/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:700 +0xea
github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).Txn(0xc001abea80, 0x5ba61c0, 0xc00b467080, 0xc00f03b9c0, 0xc00deb79b0, 0x3cab80a989)
	/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:694 +0x162
github.com/cockroachdb/cockroach/pkg/storage/reports.(*replicationConstraintStatsReportSaver).Save(0xc00b467110, 0x5ba61c0, 0xc00b467080, 0x3accfe6c, 0xed5108af7, 0x0, 0xc001abea80, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/reports/constraint_report.go:265 +0x18a
github.com/cockroachdb/cockroach/pkg/storage/reports.(*Reporter).update(0xc002bdef50, 0x5ba61c0, 0xc00b467080, 0xc00b467110, 0xc00b4670b0, 0xc00b467170, 0x5ba61c0, 0xc00b467080)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/reports/reporter.go:196 +0xb1a
github.com/cockroachdb/cockroach/pkg/storage/reports.(*Reporter).Start.func2(0x5ba61c0, 0xc00b467050)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/reports/reporter.go:122 +0x6b3
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc004d6aec0, 0xc011ee1e00, 0xc0056f4780)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:196 +0x160
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:189 +0xc4
```

cc. @andreimatei @darinpp

--- a/pkg/server/node.go
+++ b/pkg/server/node.go
@@ -32,7 +32,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlbase"
 	"github.com/cockroachdb/cockroach/pkg/storage"
 	"github.com/cockroachdb/cockroach/pkg/storage/engine"
-	"github.com/cockroachdb/cockroach/pkg/storage/reports"
 	"github.com/cockroachdb/cockroach/pkg/util"
 	"github.com/cockroachdb/cockroach/pkg/util/growstack"
 	"github.com/cockroachdb/cockroach/pkg/util/grpcutil"
@@ -143,19 +142,18 @@ func (nm nodeMetrics) callComplete(d time.Duration, pErr *roachpb.Error) {
 // IDs for bootstrapping the node itself or new stores as they're added
 // on subsequent instantiations.
 type Node struct {
-	stopper                  *stop.Stopper
-	clusterID                *base.ClusterIDContainer // UUID for Cockroach cluster
-	Descriptor               roachpb.NodeDescriptor   // Node ID, network/physical topology
-	storeCfg                 storage.StoreConfig      // Config to use and pass to stores
-	eventLogger              sql.EventLogger
-	stores                   *storage.Stores // Access to node-local stores
-	metrics                  nodeMetrics
-	recorder                 *status.MetricsRecorder
-	constraintStatsCollector *reports.Reporter
-	startedAt                int64
-	lastUp                   int64
-	initialBoot              bool // True if this is the first time this node has started.
-	txnMetrics               kv.TxnMetrics
+	stopper     *stop.Stopper
+	clusterID   *base.ClusterIDContainer // UUID for Cockroach cluster
+	Descriptor  roachpb.NodeDescriptor   // Node ID, network/physical topology
+	storeCfg    storage.StoreConfig      // Config to use and pass to stores
+	eventLogger sql.EventLogger
+	stores      *storage.Stores // Access to node-local stores
+	metrics     nodeMetrics
+	recorder    *status.MetricsRecorder
+	startedAt   int64
+	lastUp      int64
+	initialBoot bool // True if this is the first time this node has started.
+	txnMetrics  kv.TxnMetrics
 
 	perReplicaServer storage.Server
 }
@@ -273,7 +271,6 @@ func NewNode(
 		eventLogger: eventLogger,
 		clusterID:   clusterID,
 	}
-	n.constraintStatsCollector = reports.NewReporter(n.stores, &n.storeCfg)
 	n.perReplicaServer = storage.MakeServer(&n.Descriptor, n.stores)
 	return n
 }
@@ -500,8 +497,6 @@ func (n *Node) start(
 	// cluster version, but not if the server starts with a lower one and gets
 	// bumped immediately, which would be possible if gossip got started earlier).
 	n.startGossip(ctx, n.stopper)
-
-	n.constraintStatsCollector.Start(ctx, n.stopper)
 
 	allEngines := append([]engine.Engine(nil), initializedEngines...)
 	allEngines = append(allEngines, emptyEngines...)

--- a/pkg/server/server.go
+++ b/pkg/server/server.go
@@ -495,7 +495,9 @@ func NewServer(cfg Config, stopper *stop.Stopper) (*Server, error) {
 	roachpb.RegisterInternalServer(s.grpc.Server, s.node)
 	storage.RegisterPerReplicaServer(s.grpc.Server, s.node.perReplicaServer)
 	s.node.storeCfg.ClosedTimestamp.RegisterClosedTimestampServer(s.grpc.Server)
-	s.replicationReporter = reports.NewReporter(s.node.stores, &s.node.storeCfg)
+	s.replicationReporter = reports.NewReporter(
+		s.db, s.node.stores, s.storePool,
+		s.ClusterSettings(), s.nodeLiveness, internalExecutor)
 
 	s.sessionRegistry = sql.NewSessionRegistry()
 	s.jobRegistry = jobs.MakeRegistry(

--- a/pkg/server/server.go
+++ b/pkg/server/server.go
@@ -59,6 +59,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/storage/bulk"
 	"github.com/cockroachdb/cockroach/pkg/storage/closedts/container"
 	"github.com/cockroachdb/cockroach/pkg/storage/engine"
+	"github.com/cockroachdb/cockroach/pkg/storage/reports"
 	"github.com/cockroachdb/cockroach/pkg/storage/storagebase"
 	"github.com/cockroachdb/cockroach/pkg/ts"
 	"github.com/cockroachdb/cockroach/pkg/ui"
@@ -78,7 +79,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/tracing"
 	"github.com/cockroachdb/cockroach/pkg/util/uuid"
 	"github.com/cockroachdb/logtags"
-	"github.com/getsentry/raven-go"
+	raven "github.com/getsentry/raven-go"
 	gwruntime "github.com/grpc-ecosystem/grpc-gateway/runtime"
 	"github.com/opentracing/opentracing-go"
 	"github.com/pkg/errors"
@@ -184,12 +185,13 @@ type Server struct {
 	leaseMgr         *sql.LeaseManager
 	// sessionRegistry can be queried for info on running SQL sessions. It is
 	// shared between the sql.Server and the statusServer.
-	sessionRegistry    *sql.SessionRegistry
-	jobRegistry        *jobs.Registry
-	statsRefresher     *stats.Refresher
-	engines            Engines
-	internalMemMetrics sql.MemoryMetrics
-	adminMemMetrics    sql.MemoryMetrics
+	sessionRegistry     *sql.SessionRegistry
+	jobRegistry         *jobs.Registry
+	statsRefresher      *stats.Refresher
+	replicationReporter *reports.Reporter
+	engines             Engines
+	internalMemMetrics  sql.MemoryMetrics
+	adminMemMetrics     sql.MemoryMetrics
 	// sqlMemMetrics are used to track memory usage of sql sessions.
 	sqlMemMetrics sql.MemoryMetrics
 }
@@ -493,6 +495,7 @@ func NewServer(cfg Config, stopper *stop.Stopper) (*Server, error) {
 	roachpb.RegisterInternalServer(s.grpc.Server, s.node)
 	storage.RegisterPerReplicaServer(s.grpc.Server, s.node.perReplicaServer)
 	s.node.storeCfg.ClosedTimestamp.RegisterClosedTimestampServer(s.grpc.Server)
+	s.replicationReporter = reports.NewReporter(s.node.stores, &s.node.storeCfg)
 
 	s.sessionRegistry = sql.NewSessionRegistry()
 	s.jobRegistry = jobs.MakeRegistry(
@@ -1427,6 +1430,7 @@ func (s *Server) Start(ctx context.Context) error {
 		},
 		time.NewTicker,
 	)
+	s.replicationReporter.Start(ctx, s.stopper)
 
 	// Cluster ID should have been determined by this point.
 	if s.rpcContext.ClusterID.Get() == uuid.Nil {

--- a/pkg/storage/reports/reporter.go
+++ b/pkg/storage/reports/reporter.go
@@ -19,6 +19,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/keys"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/settings"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlbase"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlutil"
 	"github.com/cockroachdb/cockroach/pkg/storage"
@@ -48,8 +49,6 @@ type Reporter struct {
 	localStores *storage.Stores
 	// Constraints constructed from the locality information
 	localityConstraints []config.Constraints
-	// List of all stores known to exist on all nodes.
-	storeCfg *storage.StoreConfig
 	// The store that is the current meta 1 leaseholder
 	meta1LeaseHolder *storage.Store
 	// Latest zone config
@@ -57,7 +56,11 @@ type Reporter struct {
 	// Node liveness by store id
 	nodeLiveStatus map[roachpb.NodeID]storagepb.NodeLivenessStatus
 
-	db *client.DB
+	db        *client.DB
+	liveness  *storage.NodeLiveness
+	settings  *cluster.Settings
+	storePool *storage.StorePool
+	executor  sqlutil.InternalExecutor
 
 	frequencyMu struct {
 		syncutil.Mutex
@@ -67,11 +70,21 @@ type Reporter struct {
 }
 
 // NewReporter creates a Reporter.
-func NewReporter(localStores *storage.Stores, storeCfg *storage.StoreConfig) *Reporter {
+func NewReporter(
+	db *client.DB,
+	localStores *storage.Stores,
+	storePool *storage.StorePool,
+	st *cluster.Settings,
+	liveness *storage.NodeLiveness,
+	executor sqlutil.InternalExecutor,
+) *Reporter {
 	r := Reporter{
-		db:          storeCfg.DB,
+		db:          db,
 		localStores: localStores,
-		storeCfg:    storeCfg,
+		storePool:   storePool,
+		settings:    st,
+		liveness:    liveness,
+		executor:    executor,
 	}
 	r.frequencyMu.changeCh = make(chan struct{})
 	return &r
@@ -82,12 +95,12 @@ func NewReporter(localStores *storage.Stores, storeCfg *storage.StoreConfig) *Re
 func (stats *Reporter) reportInterval() (time.Duration, <-chan struct{}) {
 	stats.frequencyMu.Lock()
 	defer stats.frequencyMu.Unlock()
-	return ReporterInterval.Get(&stats.storeCfg.Settings.SV), stats.frequencyMu.changeCh
+	return ReporterInterval.Get(&stats.settings.SV), stats.frequencyMu.changeCh
 }
 
 // Start the periodic calls to Update().
 func (stats *Reporter) Start(ctx context.Context, stopper *stop.Stopper) {
-	ReporterInterval.SetOnChange(&stats.storeCfg.Settings.SV, func() {
+	ReporterInterval.SetOnChange(&stats.settings.SV, func() {
 		stats.frequencyMu.Lock()
 		defer stats.frequencyMu.Unlock()
 		// Signal the current waiter (if any), and prepare the channel for future
@@ -95,7 +108,7 @@ func (stats *Reporter) Start(ctx context.Context, stopper *stop.Stopper) {
 		ch := stats.frequencyMu.changeCh
 		close(ch)
 		stats.frequencyMu.changeCh = make(chan struct{})
-		stats.frequencyMu.interval = ReporterInterval.Get(&stats.storeCfg.Settings.SV)
+		stats.frequencyMu.interval = ReporterInterval.Get(&stats.settings.SV)
 	})
 	stopper.RunWorker(ctx, func(ctx context.Context) {
 		var timer timeutil.Timer
@@ -160,7 +173,7 @@ func (stats *Reporter) update(
 		log.Errorf(ctx, "unable to update the locality constraints: %s", err)
 	}
 
-	allStores := stats.storeCfg.StorePool.GetStores()
+	allStores := stats.storePool.GetStores()
 	var getStoresFromGossip StoreResolver = func(r roachpb.RangeDescriptor) []roachpb.StoreDescriptor {
 		storeDescs := make([]roachpb.StoreDescriptor, len(r.Replicas().Voters()))
 		// We'll return empty descriptors for stores that gossip doesn't have a
@@ -194,25 +207,21 @@ func (stats *Reporter) update(
 	}
 
 	if err := constraintConfVisitor.report.Save(
-		ctx, timeutil.Now() /* reportTS */, stats.db, stats.executor(),
+		ctx, timeutil.Now() /* reportTS */, stats.db, stats.executor,
 	); err != nil {
 		return errors.Wrap(err, "failed to save constraint report")
 	}
 	if err := localityStatsVisitor.report.Save(
-		ctx, timeutil.Now() /* reportTS */, stats.db, stats.executor(),
+		ctx, timeutil.Now() /* reportTS */, stats.db, stats.executor,
 	); err != nil {
 		return errors.Wrap(err, "failed to save locality report")
 	}
 	if err := statusVisitor.report.Save(
-		ctx, timeutil.Now() /* reportTS */, stats.db, stats.executor(),
+		ctx, timeutil.Now() /* reportTS */, stats.db, stats.executor,
 	); err != nil {
 		return errors.Wrap(err, "failed to save range status report")
 	}
 	return nil
-}
-
-func (stats *Reporter) executor() sqlutil.InternalExecutor {
-	return stats.storeCfg.SQLExecutor
 }
 
 // meta1LeaseHolderStore returns the node store that is the leaseholder of Meta1
@@ -241,7 +250,7 @@ func (stats *Reporter) updateLatestConfig() {
 
 func (stats *Reporter) updateLocalityConstraints() error {
 	localityConstraintsByName := make(map[string]config.Constraints, 16)
-	for _, sd := range stats.storeCfg.StorePool.GetStores() {
+	for _, sd := range stats.storePool.GetStores() {
 		c := config.Constraints{
 			Constraints: make([]config.Constraint, 0),
 		}
@@ -260,7 +269,7 @@ func (stats *Reporter) updateLocalityConstraints() error {
 }
 
 func (stats *Reporter) updateNodeLiveness() {
-	stats.nodeLiveStatus = stats.storeCfg.NodeLiveness.GetLivenessStatusMap()
+	stats.nodeLiveStatus = stats.liveness.GetLivenessStatusMap()
 }
 
 // nodeChecker checks whether a node is to be considered alive or not.

